libs/dyn/mapping.go

```diff
@@ -17,44 +19,60 @@ type Pair struct {
 // We need to use dynamic values for keys because it lets us associate metadata
 // with keys (i.e. their definition location). Keys must be strings.
 type Mapping struct {
-	pairs []Pair
-	index map[string]int
+	data         map[string]Value
+	keyLocations map[string][]Location
 }
 
 // NewMapping creates a new empty Mapping.
 func NewMapping() Mapping {
 	return Mapping{
-		pairs: make([]Pair, 0),
-		index: make(map[string]int),
+		data:         make(map[string]Value),
+		keyLocations: make(map[string][]Location),
 	}
 }
 
 // newMappingWithSize creates a new Mapping preallocated to the specified size.
 func newMappingWithSize(size int) Mapping {
 	return Mapping{
-		pairs: make([]Pair, 0, size),
-		index: make(map[string]int, size),
+		data:         make(map[string]Value, size),
+		keyLocations: make(map[string][]Location, size),
 	}
 }
 
+// NewMappingFromGoMap creates a new Mapping from a Go map of string keys and dynamic values.
+func NewMappingFromGoMap(vin map[string]Value) Mapping {
+	return newMappingFromGoMap(vin)
+}
+
 // newMappingFromGoMap creates a new Mapping from a Go map of string keys and dynamic values.
 func newMappingFromGoMap(vin map[string]Value) Mapping {
-	m := newMappingWithSize(len(vin))
-	for k, v := range vin {
-		m.Set(V(k), v) //nolint:errcheck
+	if vin == nil {
+		vin = make(map[string]Value)
+	} else {
+		vin = maps.Clone(vin)
+	}
+	return Mapping{
+		data:         vin,
+		keyLocations: make(map[string][]Location),
 	}
-	return m
 }
 
 // Pairs returns all the key-value pairs in the Mapping. The pairs are sorted by
 // their key in lexicographic order.
 func (m Mapping) Pairs() []Pair {
-	return m.pairs
+	pairs := make([]Pair, 0, len(m.data))
+	for _, k := range utils.SortedKeys(m.data) {
+		pairs = append(pairs, Pair{
+			Key:   NewValue(k, m.keyLocations[k]),
+			Value: m.data[k],
+		})
+	}
+	return pairs
```

Comment:
Curious: the iteration is identical for `Pairs/Keys/Values`. Might be a nice use case for https://go.dev/blog/range-functions#range-over-some-function-types.